syntax = "proto3";

// INTERNAL NOTES FOR MAINTAINERS:
// This protobuf package serves proxying layer for underlying services which still depend on
// protobufs from the `azimuth.rules.v1` package. Any updates to this file need to be ported
// in copy-paste fashion to `azimuth.rules.v1`.
//
// Also, be sure to update the following Go package to convert between the new and legacy types.:
// - azimuth.services.repo.rule.v1.pbmapper
package sift.rules.v1;

import "google/api/annotations.proto";
import "google/api/field_behavior.proto";
import "google/protobuf/timestamp.proto";
import "protoc-gen-openapiv2/options/annotations.proto";
import "sift/annotations/v1/annotations.proto";


option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_swagger) = {
  info: {title: "Rule Service"}
};

service RuleService {
  option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_tag) = {
    description: "Service to programmatically interact with rules."
    external_docs: {
      url: "https://customer.support.siftstack.com/servicedesk/customer/portal/2/article/265421102";
      description: "Read more about what rules are.";
    }
  };

  // Queries rules based on provided search parameters.
  rpc SearchRules(SearchRulesRequest) returns (SearchRulesResponse) {
    option (google.api.http) = {
      post: "/api/v1/rules/search"
      body: "*"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "SearchRules"
      description: "Queries rules based on provided search parameters."
    };
  }

  // Retrieves the latest version of a rule.
  rpc GetRule(GetRuleRequest) returns (GetRuleResponse) {
    option (google.api.http) = {get: "/api/v1/rules"};
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "GetRule"
      description: "Retrieves the latest version of a rule."
    };
  }

  // Retrieve multiple rules.
  rpc BatchGetRules(BatchGetRulesRequest) returns (BatchGetRulesResponse) {
    option (google.api.http) = {
      post: "/api/v1/rules:batchGet"
      body: "*"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "BatchGetRules"
      description: "Retrieves multiple rules."
    };
  }

  // Creates a rule.
  rpc CreateRule(CreateRuleRequest) returns (CreateRuleResponse) {
    option (google.api.http) = {
      post: "/api/v1/rules"
      body: "*"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "CreateRule"
      description: "Creates a rule."
    };
  }

  // Updates an existing rule.
  rpc UpdateRule(UpdateRuleRequest) returns (UpdateRuleResponse) {
    option (google.api.http) = {
      put: "/api/v1/rules"
      body: "*"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "UpdateRule"
      description: "Updates an existing rule."
    };
  }

  // Deletes a rule
  rpc DeleteRule(DeleteRuleRequest) returns (DeleteRuleResponse) {
    option (google.api.http) = {
      post: "/api/v1/rules/delete"
      body: "*"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "DeleteRule"
      description: "Deletes a rule."
    };
  }

  // Evaluates the provided rules and generate annotations based on the result.
  rpc EvaluateRules(EvaluateRulesRequest) returns (EvaluateRulesResponse) {
    option (google.api.http) = {
      post: "/api/v1/rules/evaluate"
      body: "*"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "EvaluateRules"
      description: "Evaluates the provided rules and generate annotations based on the result."
    };
  }

  // Deprecated - use ViewJsonRules instead. Retrieve a JSON object containing all of the rules for a given asset.
  rpc ViewHumanFriendlyRules(ViewHumanFriendlyRulesRequest) returns (ViewHumanFriendlyRulesResponse) {
    option deprecated = true;
    option (google.api.http) = {get: "/api/v1/rules:viewHumanFriendly"};
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "ViewHumanFriendlyRules"
      description: "Retrieve a JSON object containing all of the rules for a given asset."
    };
  }

  // Retrieve a JSON object containing all of the rules for a given asset.
  rpc ViewJsonRules(ViewJsonRulesRequest) returns (ViewJsonRulesResponse) {
    option (google.api.http) = {get: "/api/v1/rules:viewJson"};
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "ViewJsonRules"
      description: "Retrieve a JSON object containing all of the rules for a given asset."
    };
  }

  // Deprecated - use UpdateJsonRules instead. Batch update rules given the `rules_json` which is a JSON list of rules.
  rpc UpdateHumanFriendlyRules(UpdateHumanFriendlyRulesRequest) returns (UpdateHumanFriendlyRulesResponse) {
    option deprecated = true;
    option (google.api.http) = {
      post: "/api/v1/rules:updateHumanFriendly"
      body: "*"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "UpdateHumanFriendlyRules"
      description: "Batch update rules given the `rules_json` which is a JSON list of rules."
    };
  }

  // Validate a batch update for rules given the `rules_json` which is a JSON list of rules. This is a dry-run operation.
  rpc ValidateJsonRules(ValidateJsonRulesRequest) returns (ValidateJsonRulesResponse) {
    option (google.api.http) = {
      post: "/api/v1/rules:validateJson"
      body: "*"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "ValidateJsonRules"
      description: "Validate a batch update for rules given the `rules_json` which is a JSON list of rules."
    };
  }

  // Batch update rules given the `rules_json` which is a JSON list of rules.
  rpc UpdateJsonRules(UpdateJsonRulesRequest) returns (UpdateJsonRulesResponse) {
    option (google.api.http) = {
      post: "/api/v1/rules:updateJson"
      body: "*"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "UpdateJsonRules"
      description: "Batch update rules given the `rules_json` which is a JSON list of rules."
    };
  }

  // Retrieves a list of rule versions for the given rule.
  rpc ListRuleVersions(ListRuleVersionsRequest) returns (ListRuleVersionsResponse) {
    option (google.api.http) = {get: "/api/v1/rules/{rule_id}/versions"};
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "ListRuleVersions"
      description: "Retrieves a list of rule versions for the given rule."
    };
  }

  // Retrieves a specific version of a rule.
  rpc GetRuleVersion(GetRuleVersionRequest) returns (GetRuleVersionResponse) {
    option (google.api.http) = {get: "/api/v1/rules:version"};
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "GetRuleVersion"
      description: "Retrieves a specific version of a rule."
    };
  }

  // Retrieves multiple rules by rule versions.
  rpc BatchGetRuleVersions(BatchGetRuleVersionsRequest) returns (BatchGetRuleVersionsResponse) {
    option (google.api.http) = {
      post: "/api/v1/rules:batchGetVersions"
      body: "*"
    };
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
      summary: "BatchGetRuleVersions"
      description: "Retrieves multiple rules by rule versions."
    };
  }
}

message Rule {
  // 5 is reserved from old "current_status" column directly on rule
  reserved 5;
  string rule_id = 1 [(google.api.field_behavior) = REQUIRED];
  string asset_id = 2 [(google.api.field_behavior) = REQUIRED];
  string name = 3 [(google.api.field_behavior) = REQUIRED];
  string description = 4 [(google.api.field_behavior) = REQUIRED];
  bool is_enabled = 6 [(google.api.field_behavior) = REQUIRED];
  google.protobuf.Timestamp created_date = 7 [(google.api.field_behavior) = REQUIRED];
  google.protobuf.Timestamp modified_date = 8 [(google.api.field_behavior) = REQUIRED];
  string created_by_user_id = 9 [(google.api.field_behavior) = REQUIRED];
  string modified_by_user_id = 10 [(google.api.field_behavior) = REQUIRED];
  string organization_id = 11 [(google.api.field_behavior) = REQUIRED];
  repeated RuleCondition conditions = 12 [(google.api.field_behavior) = REQUIRED];
  RuleVersion rule_version = 13 [(google.api.field_behavior) = REQUIRED];
}
message RuleCondition {
  // 4 is reserved from old "status" column that lived directly on rule condition
  reserved 4;
  string rule_condition_id = 1 [(google.api.field_behavior) = REQUIRED];
  string rule_id = 2 [(google.api.field_behavior) = REQUIRED];
  RuleConditionExpression expression = 3 [(google.api.field_behavior) = REQUIRED];
  google.protobuf.Timestamp created_date = 5 [(google.api.field_behavior) = REQUIRED];
  google.protobuf.Timestamp modified_date = 6 [(google.api.field_behavior) = REQUIRED];
  string created_by_user_id = 7 [(google.api.field_behavior) = REQUIRED];
  string modified_by_user_id = 8 [(google.api.field_behavior) = REQUIRED];
  repeated RuleAction actions = 9 [(google.api.field_behavior) = REQUIRED];
  string rule_condition_version_id = 10 [(google.api.field_behavior) = REQUIRED];
}
message RuleAction {
  string rule_action_id = 1 [(google.api.field_behavior) = REQUIRED];
  string rule_condition_id = 2 [(google.api.field_behavior) = REQUIRED];
  ActionKind action_type = 3 [(google.api.field_behavior) = REQUIRED];
  RuleActionConfiguration configuration = 4 [(google.api.field_behavior) = REQUIRED];
  google.protobuf.Timestamp created_date = 5 [(google.api.field_behavior) = REQUIRED];
  google.protobuf.Timestamp modified_date = 6 [(google.api.field_behavior) = REQUIRED];
  string created_by_user_id = 7 [(google.api.field_behavior) = REQUIRED];
  string modified_by_user_id = 8 [(google.api.field_behavior) = REQUIRED];
  string rule_action_version_id = 9 [(google.api.field_behavior) = REQUIRED];
}

message RuleAssetConfiguration {
  repeated string asset_ids = 1;
}

message SearchRulesRequest {
  // Max number of rules to return (returns all if omitted).
  optional uint32 limit = 1;
  // Only applies if limit provided.
  uint32 offset = 2;
  // Order to sort results by (defaults to ascending).
  optional SearchOrder order = 3;
  string name_matches = 4;
  bool case_sensitive = 5;
  bool regexp = 6;
  optional string order_by = 7;
  // If provided, only returns rules with the given ids
  repeated string rule_ids = 8;
  // If provided, only returns rules with the given asset ids
  repeated string asset_ids = 9;
}
enum SearchOrder {
  SEARCH_ORDER_UNSPECIFIED = 0 [deprecated = true];
  SEARCH_ORDER_ASC = 1;
  SEARCH_ORDER_DESC = 2;
}
message SearchRulesResponse {
  uint32 count = 1 [(google.api.field_behavior) = REQUIRED];
  // Conditions are not included in the search response. The latest version of the rule is returned.
  repeated Rule rules = 2 [(google.api.field_behavior) = REQUIRED];
}

message GetRuleRequest {
  string rule_id = 1 [(google.api.field_behavior) = REQUIRED];
}
message GetRuleResponse {
  Rule rule = 1 [(google.api.field_behavior) = REQUIRED];
}

message BatchGetRulesRequest {
  repeated string rule_ids = 1 [(google.api.field_behavior) = REQUIRED];
}
message BatchGetRulesResponse {
  repeated Rule rules = 1 [(google.api.field_behavior) = REQUIRED];
}

message CreateRuleRequest {
  UpdateRuleRequest update = 1 [(google.api.field_behavior) = REQUIRED];
}
message CreateRuleResponse {
  string rule_id = 1 [(google.api.field_behavior) = REQUIRED];
}
message UpdateRuleRequest {
  optional string rule_id = 1;
  string name = 2 [(google.api.field_behavior) = REQUIRED];
  string description = 3 [(google.api.field_behavior) = REQUIRED];
  string asset_id = 4 [(google.api.field_behavior) = REQUIRED];
  bool is_enabled = 5 [(google.api.field_behavior) = REQUIRED];
  repeated UpdateConditionRequest conditions = 6 [(google.api.field_behavior) = REQUIRED];
  string organization_id = 7;
  string version_notes = 8;
}
message UpdateConditionRequest {
  // 2 is reserved from old status update support on rule condition directly
  reserved 2;
  optional string rule_condition_id = 1;
  RuleConditionExpression expression = 3 [(google.api.field_behavior) = REQUIRED];
  repeated UpdateActionRequest actions = 4 [(google.api.field_behavior) = REQUIRED];
}
message UpdateActionRequest {
  optional string rule_action_id = 1;
  ActionKind action_type = 2 [(google.api.field_behavior) = REQUIRED];
  RuleActionConfiguration configuration = 3 [(google.api.field_behavior) = REQUIRED];
}
enum ActionKind {
  ACTION_KIND_UNSPECIFIED = 0 [deprecated = true];
  NOTIFICATION = 1;
  ANNOTATION = 2;
}
message UpdateRuleResponse {
  string rule_id = 1 [(google.api.field_behavior) = REQUIRED];
}

message DeleteRuleRequest {
  string rule_id = 1 [(google.api.field_behavior) = REQUIRED];
}
message DeleteRuleResponse {}

// Deprecated - use ViewJsonRulesRequest.
message ViewHumanFriendlyRulesRequest {
  option deprecated = true;
  string asset_id = 1 [(google.api.field_behavior) = REQUIRED];
}

// Deprecated - use ViewJsonRulesResponse.
message ViewHumanFriendlyRulesResponse {
  option deprecated = true;
  string rules_json = 1 [(google.api.field_behavior) = REQUIRED];
}

// Deprecated - use UpdateJsonRulesRequest.
message UpdateHumanFriendlyRulesRequest {
  option deprecated = true;
  string asset_id = 1 [(google.api.field_behavior) = REQUIRED];
  string rules_json = 2 [(google.api.field_behavior) = REQUIRED];
  string organization_id = 3 [(google.api.field_behavior) = REQUIRED];
}

// Deprecated - use UpdateJsonRulesResponse.
message UpdateHumanFriendlyRulesResponse {
  option deprecated = true;
  bool success = 1 [(google.api.field_behavior) = REQUIRED];
  int32 rules_count = 2 [(google.api.field_behavior) = REQUIRED];
  string messages = 3 [(google.api.field_behavior) = REQUIRED];
}

message ViewJsonRulesRequest {
  string asset_id = 1 [(google.api.field_behavior) = REQUIRED];
}

message ViewJsonRulesResponse {
  string rules_json = 1 [(google.api.field_behavior) = REQUIRED];
}

message JsonRulesRequest {
  string asset_id = 1 [(google.api.field_behavior) = REQUIRED];
  string rules_json = 2 [(google.api.field_behavior) = REQUIRED];
  string organization_id = 3 [(google.api.field_behavior) = REQUIRED];
}

message JsonRulesResponse {
  bool success = 1 [(google.api.field_behavior) = REQUIRED];
  int32 total_rules_count = 2 [(google.api.field_behavior) = REQUIRED];
  int32 rules_created_count = 3 [(google.api.field_behavior) = REQUIRED];
  int32 rules_updated_count = 4 [(google.api.field_behavior) = REQUIRED];
  int32 rules_deleted_count = 5 [(google.api.field_behavior) = REQUIRED];
  optional string error_messages = 6;
}

message ValidateJsonRulesRequest {
  JsonRulesRequest request = 1 [(google.api.field_behavior) = REQUIRED];
}

message ValidateJsonRulesResponse {
  JsonRulesResponse response = 1 [(google.api.field_behavior) = REQUIRED];
}

message UpdateJsonRulesRequest {
  JsonRulesRequest request = 1 [(google.api.field_behavior) = REQUIRED];
}

message UpdateJsonRulesResponse {
  JsonRulesResponse response = 1 [(google.api.field_behavior) = REQUIRED];
}

message ListRuleVersionsRequest {
  string rule_id = 1 [(google.api.field_behavior) = REQUIRED];

  // The maximum number of Rule Versions to return.
  // The service may return fewer than this value.
  // If unspecified, at most 50 Rule Versions will be returned.
  // The maximum value is 1000; values above 1000 will be coerced to 1000.
  uint32 page_size = 2 [(google.api.field_behavior) = OPTIONAL];

  // A page token, received from a previous `ListRuleVersions` call.
  // Provide this to retrieve the subsequent page.
  // When paginating, all other parameters provided to `ListRuleVersions` must match
  // the call that provided the page token.
  string page_token = 3 [(google.api.field_behavior) = OPTIONAL];

  // A [Common Expression Language (CEL)](https://github.com/google/cel-spec) filter string.
  // Available fields to filter by are `rule_version_id`, `user_notes`,  and `change_message`.
  // For further information about how to use CELs, please refer to [this guide](https://github.com/google/cel-spec/blob/master/doc/langdef.md#standard-definitions). Optional.
  string filter = 4;
}

message RuleVersion {
  string rule_id = 1 [(google.api.field_behavior) = REQUIRED];
  string rule_version_id = 2 [(google.api.field_behavior) = REQUIRED];
  string version = 3 [(google.api.field_behavior) = REQUIRED];
  google.protobuf.Timestamp created_date = 4 [(google.api.field_behavior) = REQUIRED];
  string created_by_user_id = 5 [(google.api.field_behavior) = REQUIRED];
  string version_notes = 6 [(google.api.field_behavior) = REQUIRED];
  string generated_change_message = 7 [(google.api.field_behavior) = REQUIRED];
}

message ListRuleVersionsResponse {
  repeated RuleVersion rule_versions = 1 [(google.api.field_behavior) = REQUIRED];
  string next_page_token = 2;
}

message GetRuleVersionRequest {
  string rule_version_id = 1 [(google.api.field_behavior) = REQUIRED];
}

message GetRuleVersionResponse {
  Rule rule = 1 [(google.api.field_behavior) = REQUIRED];
}

message BatchGetRuleVersionsRequest {
  repeated string rule_version_ids = 1 [(google.api.field_behavior) = REQUIRED];
}

message BatchGetRuleVersionsResponse {
  repeated Rule rules = 1 [(google.api.field_behavior) = REQUIRED];
}

message RuleConditionExpression {
  oneof expression {
    SingleChannelComparisonExpression single_channel_comparison = 1 [deprecated = true];
    CalculatedChannelConfig calculated_channel = 2;
  }
}

// Deprecated - use CalculatedChannelConfig.
message SingleChannelComparisonExpression {
  string channel_component = 1 [(google.api.field_behavior) = REQUIRED];
  string channel_name = 2 [(google.api.field_behavior) = REQUIRED];
  ConditionComparator comparator = 3 [(google.api.field_behavior) = REQUIRED];
  // Threshold can be either a double or a string. Boolean values are encoded as 1 or 0. Enum values are stored as the string representation.
  oneof threshold {
    double double = 4;
    string string = 5;
    LastValueThreshold last_value = 6;
  }
}
message LastValueThreshold {}
enum ConditionComparator {
  CONDITION_COMPARATOR_UNSPECIFIED = 0 [deprecated = true];
  LESS_THAN = 1;
  LESS_THAN_OR_EQUAL = 2;
  GREATER_THAN = 3;
  GREATER_THAN_OR_EQUAL = 4;
  EQUAL = 5;
  NOT_EQUAL = 6;
}
message CalculatedChannelConfig {
  map<string, ChannelReference> channel_references = 1 [(google.api.field_behavior) = REQUIRED];
  string expression = 2 [(google.api.field_behavior) = REQUIRED];
}
message ChannelReference {
  string name = 1 [(google.api.field_behavior) = REQUIRED];
  string component = 2 [(google.api.field_behavior) = REQUIRED];
}

message RuleActionConfiguration {
  oneof configuration {
    NotificationActionConfiguration notification = 1;
    AnnotationActionConfiguration annotation = 2;
  }
}
message NotificationActionConfiguration {
  repeated string recipient_user_ids = 1 [(google.api.field_behavior) = REQUIRED];
}
message AnnotationActionConfiguration {
  repeated string tag_ids = 1 [(google.api.field_behavior) = REQUIRED];
  sift.annotations.v1.AnnotationType annotation_type = 2;
  optional string assigned_to_user_id = 3;
}

message EvaluateRulesRequest {
  repeated string rule_ids = 1 [(google.api.field_behavior) = REQUIRED];
  EvaluatedAnnotationOptions annotation_options = 2;
  oneof time {
    string run_id = 3;
    TimeRangeQuery time_range = 4;
  }
  bool dry_run = 5;
}
message EvaluatedAnnotationOptions {
  repeated string tag_ids = 1;
}
message TimeRangeQuery {
  google.protobuf.Timestamp start_time = 1;
  google.protobuf.Timestamp end_time = 2;
}

message EvaluateRulesResponse {
  int32 created_annotation_count = 1;
  // If dry_run is true, this will be populated with the annotations that would be created
  repeated DryRunAnnotation dry_run_annotations = 2;
}

message DryRunAnnotation {
  string condition_id = 1;
  string name = 2;
  google.protobuf.Timestamp start_time = 3;
  google.protobuf.Timestamp end_time = 4;
  string condition_version_id = 5;
}
